### ISSUE TITLE
File Transfer

### ISSUE THREAD

**Issue by onimor** [grpc/grpc-dotnet#1467]
Blazor WASM, .NET5
I'm trying to transfer files to the server and back. I transmit parts of 3mb in the form of bytes. This consumes a lot of RAM by the browser. And when I transfer from the server, it happens very slowly for files larger than 3mb. Maybe there is a way to transfer a stream, not a byte? or maybe I have a failed approach?

proto

```
service Greeter {
  rpc UploadFile(stream FileRequest) returns (StatusRequest);
  rpc DownloadFile(FilesInfoRequest) returns (stream FileRequest)
}
message StatusRequest {
  string Status = 1;
}
message FileRequest {
    string FileName = 1;
    bytes FileBytes =2;
   
}
message FilesInfoRequest {
    string FileName = 1;
    string FilePath = 2;   
}
```

Sending method. Executed in the browser. Works fast.
Google.Protobuf.UnsafeByteOperations.UnsafeWrap  -  consumes a lot of RAM ~ file 120mb - 1 gigabytes
```
private const int _maxSize = 3 * 1024 * 1024;
private byte[] _part = new byte[_maxSize];
private FileRequest _fileRequest;
private async void UploadFiles(InputFileChangeEventArgs e)
    {
      
        foreach (var file in e.GetMultipleFiles())
        {
            try
            {
                using var call = GreeterClient.UploadFile();
                using (var ms = file.OpenReadStream(file.Size))
                {
                   
                    while (ms.Length - ms.Position > 0)
                    {
                        if (ms.Length - ms.Position <= _maxSize)
                            _part = new byte[ms.Length - ms.Position];

                        await ms.ReadAsync(_part, 0, _part.Length);

                        _fileRequest = new FileRequest { FileName = file.Name, FileBytes = Google.Protobuf.UnsafeByteOperations.UnsafeWrap(_part) };
                        await call.RequestStream.WriteAsync(_fileRequest);
                        _fileRequest = null; 
                        
                    }
                }
                await call.RequestStream.CompleteAsync();
                var response = await call;
                var status = response.Status;
            }
            catch(Exception ex)
            {
                ShowMessage(ex.Message, Severity.Error);
            }
        }
    }
```

Method of getting the file from the server (download). Running on the server. It works very slowly. Almost does not consume RAM.

```
private const int _maxSize = 3 * 1024 * 1024;
        private byte[] _part = new byte[_maxSize];
        public override async Task DownloadFile(FilesInfoRequest request, IServerStreamWriter<FileRequest> responseStream, ServerCallContext context)
        {
            using (FileStream fstream = new FileStream(request.FilePath,FileMode.Open))
            {
                while (fstream.Length - fstream.Position > 0)
                {
                    if (fstream.Length - fstream.Position <= _maxSize)
                        _part = new byte[fstream.Length - fstream.Position];

                    await fstream.ReadAsync(_part, 0, _part.Length);

                    await responseStream.WriteAsync(new FileRequest
                    {
                        FileName = fstream.Name,
                        FileBytes = UnsafeByteOperations.UnsafeWrap(_part)
                    });

                }

                   
            }
}
```

What's going wrong?

**Comment by JamesNK** [comment]
> rpc UploadFile(stream FileRequest) returns (StatusRequest);

I'm surprised this doesn't error. gRPC-Web in Blazor WebAssembly doesn't support client streaming. If it does work, then the client must be buffering all data before sending it. That would explain why you have high memory usage.

I'm not sure about the download slowness. Is this between apps on the same computer, or is there latency involved?

**Comment by onimor** [comment]
@JamesNK , wow ... how, then, to send files to the server? 

this is for 1 computer, 1 solution. there is no delay. but a 100 mega byte file takes about 3 minutes...

**Comment by JamesNK** [comment]
I don't know. I'm not a browser expert at this. As far as I know, the only HTTP API available in JS and WASM is fetch, and fetch doesn't support streaming uploads. You're better off uploading large files from the browser.

You could ask at dotnet/aspnetcore for a more expert opinion.

I'll look into slow download and see if I can recreate it.

**Comment by onimor** [comment]
@JamesNK  Thanks, I have a night now. I think in the afternoon I could make a separate project for you, so that it would be easier to check.

**Comment by onimor** [comment]
@JamesNK  [https://github.com/onimor/CIMFileTest](Here is a test project with download and upload)
Just add your file larger than 100 mb via the page on the website or along the way: CIMFileTest/Backend/CIM.Backend/Resources/FileTest

**Comment by captainsafia** [comment]
@onimor Is the guidance under https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=webassembly helpful?

**Comment by onimor** [comment]
@captainsafia , did as described in the end I get the same large memory consumption.

**Comment by captainsafia** [comment]
> @captainsafia , did as described in the end I get the same large memory consumption. You can run the project to which I made a link above and check it out to make sure.

I noticed you have an upload implementation that uses gRPC and one that doesn't. Are you experiencing the large memory consumption in both?

I suspect that you're simply hitting the limits of the `InputFile` component by processing 100MB files on the browser. As mentioned above, I'd recommend opening an issue on the dotnet/aspnetcore repo for specific guidance from the Blazor team.

**Comment by onimor** [comment]
@captainsafia, Thanks for the feedback. There are actually 2 methods. Initially there was gRPC, after many tips with reference to the example from Microsoft, I made another 1. But memory consumption is no different in both cases. I wrote to asp, they referred to the same example and closed the question, without even checking the problem....
[Here is my question to them](https://github.com/dotnet/aspnetcore/issues/38166)
